product_docs/docs/pgd/5/parallelapply.mdx

```diff
@@ -5,16 +5,26 @@ navTitle: Parallel Apply
+!!!
+
 Parallel Apply is a feature of PGD that allows a PGD node to use multiple writers per subscription. This behavior generally increases the throughput of a subscription and improves replication performance.
 
-The transactional changes from the subscription are written by the multiple Parallel Apply writers. However, each writer ensures that the final commit of its transaction doesn't violate the commit order as executed on the origin node. If there's a violation, an error occurs, and the transaction can be rolled back. 
+When Parallel Apply is operating, the transactional changes from the subscription are written by multiple writers. However, each writer ensures that the final commit of its transaction doesn't violate the commit order as executed on the origin node. If there's a violation, an error is generated and the transaction can be rolled back. 
 
-!!! Warning Possible deadlocks
-It might be possible for this out-of-order application of changes to trigger a deadlock. PGD currently resolves such deadlocks between Parallel Apply writers by aborting and retrying the transactions involved. Experiencing a large number of such deadlocks is an indication that Parallel Apply isn't a good fit for your workload. In this case, consider disabling it.
-!!!
+This mechanism has previously meant that if a transaction is pending commit and modifies a row that another transaction needs to change, and that other transaction executed on the origin node before the pending transaction did, then the resulting error could manifest itself as a deadlock if the pending transaction has taken out a lock request. 
+
+Additionally, handling the error could increase replication lag, due to a combination of the time taken to detect the deadlock, the time taken for the client to rollback its transaction, the time taken for indirect garbage collection of the already applied changes, and the time taken to redo the work. 
+
+This is where Parallel Apply’s deadlock mitigation, introduced in PGD 5.2, comes into play.
 
-## Configuring Parallel Apply
-Two variables control Parallel Apply in PGD 5: [`bdr.max_writers_per_subscription`](/pgd/latest/reference/pgd-settings#bdrmax_writers_per_subscription) (defaults to 8) and [`bdr.writers_per_subscription`](/pgd/latest/reference/pgd-settings#bdrwriters_per_subscription) (defaults to 2).
+Parallel Apply’s deadlock mitigation is always enabled. Parallel Apply deadlock mitigation gets transactions to preview their write sets before they are executed. Transactions can then probe other contemporary transactions with older commit timestamps looking for a write set intersection. 
+
+Where there are intersections, Parallel Apply executes the transaction until it finds a conflicting row. Only then will it put the transaction on hold until the intersecting transactions have committed the conflicted rows. It then continues applying row changes from the transaction.
+
```

Comment:
I agree with PM's input this morning regarding implementation details not being part of external doc. IMHO, it's not so much that IP or trade secrets might be revealed to competitors as it is keeping it simple to describe the value proposition how to put the feature to use.

OTOH, PA deadlock mitigation is something to highlight in the 5.2 release notes. Many users may have put the feature aside because it did not work for them; they should be encouraged to give it another try.

Reply:
Taking content out at this point is problematic because there's no concept documentation to fall back to. And this is a detail, which as you point out fixes an existing issue which may have put users aside.

The current release notes format isn't built for anything more than a cursory highlight and I would suggest, at least for 5.2 that we leave this in place till we have concept documentation that can allow us the cut down on the implementation detail.

Reply:
I sat down and reduced the detail in the deadlock mitigation to a simple "It sees a conflict, it puts it on hold." - This should suffice for the user manual.